core/src/main/scala/org/http4s/Message.scala

```diff
@@ -373,38 +373,39 @@ final class Request[F[_]](
 
   private def connectionInfo: Option[Connection] = attributes.lookup(Keys.ConnectionInfo)
 
-  def remote: Option[InetSocketAddress] = connectionInfo.map(_.remote)
+  def remote: Option[SocketAddress[IpAddress]] = connectionInfo.map(_.remote)
 
   /** Returns the the X-Forwarded-For value if present, else the remote address.
     */
-  def from: Option[InetAddress] =
+  def from: Option[IpAddress] =
     headers
       .get(`X-Forwarded-For`)
-      .fold(remote.flatMap(remote => Option(remote.getAddress)))(_.values.head)
+      .fold(remote.map(_.ip))(_.values.head)
 
-  def remoteAddr: Option[String] = remote.map(_.getHostString)
+  def remoteAddr: Option[IpAddress] = remote.map(_.ip)
 
-  def remoteHost: Option[String] = remote.map(_.getHostName)
+  def remoteHost(implicit F: Sync[F]): F[Option[Hostname]] = {
+    val inetAddress = remote.map(_.ip.toInetAddress)
+    F.delay(inetAddress.map(_.getHostName)).map(_.flatMap(Hostname(_)))
```

Comment:
I think this should have been getHostAddress all along, and then it wouldn't need the `F`?  Why are we doing reverse lookups here?